// @atlaskit/button/dist/esm/performance/examples.js
import { __awaiter, __generator, __read } from "tslib";
import React, { useLayoutEffect, useRef, useState } from 'react';
import Button, { CustomThemeButton } from '../index';
// eslint-disable-next-line
import { fireEvent } from '@testing-library/react';
function Example(_a) {
    var Component = _a.Component;
    var _b = __read(useState(false), 2), isSelected = _b[0], setIsSelected = _b[1];
    var _c = __read(useState(false), 2), isDisabled = _c[0], setIsDisabled = _c[1];
    var ref = useRef(null);
    useLayoutEffect(function () {
        function toggleSelect() {
            setIsSelected(function (value) { return !value; });
        }
        function toggleDisabled() {
            setIsDisabled(function (value) { return !value; });
        }
        var el = ref.current;
        if (!el) {
            throw new Error('Could not find button ref');
        }
        el.addEventListener('toggle-select', toggleSelect);
        el.addEventListener('toggle-disabled', toggleDisabled);
        return function () {
            el.removeEventListener('toggle-select', toggleSelect);
            el.removeEventListener('toggle-disabled', toggleDisabled);
        };
    }, []);
    return (
    // @ts-ignore
    React.createElement(Component, { ref: ref, testId: "my-button", isSelected: isSelected, isDisabled: isDisabled, "data-is-selected": isSelected }, "Hello world"));
}
var interactionTasks = [
    {
        name: 'Click a button',
        description: 'Recording how long a mousedown + click event take to be processed',
        run: function (_a) {
            var container = _a.container;
            return __awaiter(void 0, void 0, void 0, function () {
                var button;
                return __generator(this, function (_b) {
                    button = container.querySelector('[data-testid="my-button"]');
                    if (button == null) {
                        throw new Error('Could not find button element');
                    }
                    fireEvent.mouseDown(button);
                    fireEvent.click(button);
                    return [2 /*return*/];
                });
            });
        },
    },
    {
        name: 'Focus on button',
        description: 'Focus on a button and wait for layout and paint to finish',
        run: function (_a) {
            var container = _a.container, controls = _a.controls;
            return __awaiter(void 0, void 0, void 0, function () {
                var button;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            button = container.querySelector('[data-testid="my-button"]');
                            if (button == null) {
                                throw new Error('Could not find button element');
                            }
                            return [4 /*yield*/, controls.time(function () { return __awaiter(void 0, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        fireEvent.focus(button);
                                        fireEvent.blur(button);
                                        return [2 /*return*/];
                                    });
                                }); })];
                        case 1:
                            _b.sent();
                            return [2 /*return*/];
                    }
                });
            });
        },
    },
    {
        name: 'Select button',
        description: 'Trigger the selection of a button',
        run: function (_a) {
            var container = _a.container, controls = _a.controls;
            return __awaiter(void 0, void 0, void 0, function () {
                var button;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            button = container.querySelector('[data-testid="my-button"]');
                            if (button == null) {
                                throw new Error('Could not find button element');
                            }
                            if (button.getAttribute('data-is-selected') !== 'false') {
                                throw new Error('Should not start selected');
                            }
                            return [4 /*yield*/, controls.time(function () { return __awaiter(void 0, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        fireEvent(button, new Event('toggle-select'));
                                        return [2 /*return*/];
                                    });
                                }); })];
                        case 1:
                            _b.sent();
                            if (button.getAttribute('data-is-selected') !== 'true') {
                                throw new Error('Should now be selected');
                            }
                            return [2 /*return*/];
                    }
                });
            });
        },
    },
    {
        name: 'Disable button',
        description: 'Trigger the disabling of a button',
        run: function (_a) {
            var container = _a.container, controls = _a.controls;
            return __awaiter(void 0, void 0, void 0, function () {
                var button;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            button = container.querySelector('[data-testid="my-button"]');
                            if (button == null) {
                                throw new Error('Could not find button element');
                            }
                            if (button.hasAttribute('disabled')) {
                                throw new Error('Should not start disabled');
                            }
                            return [4 /*yield*/, controls.time(function () { return __awaiter(void 0, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        fireEvent(button, new Event('toggle-disabled'));
                                        return [2 /*return*/];
                                    });
                                }); })];
                        case 1:
                            _b.sent();
                            if (!button.hasAttribute('disabled')) {
                                throw new Error('Should now be disabled');
                            }
                            return [2 /*return*/];
                    }
                });
            });
        },
    },
];
export var button = function () { return React.createElement(Example, { Component: Button }); };
button.story = {
    name: 'Button',
    parameters: {
        performance: {
            interactions: interactionTasks,
        },
    },
};
export var customThemeButton = function () { return (React.createElement(Example, { Component: CustomThemeButton })); };
customThemeButton.story = {
    name: 'CustomThemeButton',
    parameters: {
        performance: {
            interactions: interactionTasks,
        },
    },
};
